### PR TITLE
Fix PulseQobj converter name collision

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary

This PR fixes name collision bug (edge case) of pulse qobj converter (`QobjToInstructionConverter`).

### Details and comments

In the pulse qobj model, instruction is managed by the unique string identifier, e.g. `ShiftPhase`: `fc`, in the transfer layer. Only exception is the `Play` instruction, because this instruction takes two types of operands, namely, parametric pulse and waveform. When the backend doesn't support a parametric pulse form, this is converted into waveform and actual waveform samples are stored in the `.pulse_library` field. Thus command `.name` of the play instruction is `parametric_pulse` or unique waveform name (such as `CX_d10_u15`, which is a foreign key to refer to the pulse library), rather than "play".

Currently, `QobjToInstructionConverter` manages these instructions with a dedicated method. This is sort of descriptor pattern, where dedicated method to parse each pulse library item (as well as other instructions) is dynamically generated. Note that this descriptor is directly attached to the class. Thus, when multiple converters are instantiated with different backends in the same thread, it could cause name collision when backend reports different waveforms under the same name. For example,

```python
pulse_library_from_backend_x = [PulseLibraryItem(name="pulse123", samples=[0.1, 0.1, 0.1])]
converter_of_backend_x = QobjToInstructionConverter(pulse_library_from_backend_x, buffer=0)

pulse_library_from_backend_y = [PulseLibraryItem(name="pulse123", samples=[0.2, 0.2, 0.2])]
converter_of_backend_y = QobjToInstructionConverter(pulse_library_from_backend_y, buffer=0)

qobj1 = PulseQobjInstruction(name="pulse123", qubits=[0], t0=0, ch="d0")

sched_out_x = converter_of_backend_x(qobj1)
sched_out_y = converter_of_backend_y(qobj1)
```

`sched_out_x` and `sched_out_y` accidentally loads the same pulse from the pulse library of backend y (because this is the latest instance). Note that this doesn't happen in current Qiskit client, because converter is immediately discarded after `PulseDefaults` for a particular backend is instantiated. However, this mechanism doesn't stop developers to keep converter. Such converter is silently mutated when another backend is loaded.

In this PR, descriptor-like pattern is replaced with visitor pattern, and class level cache is removed.

New test `test_instruction_name_collision` is also added.

### CODE REVIEW

**Comment by nkanazawa1989** [comment]
I confirmed new mechanism works properly with real backend.

```python
from qiskit import QuantumCircuit, transpile, schedule

qc = QuantumCircuit(2)
qc.h(0)
qc.cx(0, 1)
qc.measure_all()

# chose qubit pair with DDCX,
# this is currently waveform instruction because of missing parametric pulse class.
# this is nice to debug pulse library reference mechanism.
bell_schedule = schedule(transpile(qc, backend, initial_layout=[3, 5]), backend)
job = backend.run(bell_schedule)
result = job.result()
result.get_counts()  # {'00': 1929, '01': 557, '10': 88, '11': 1426}
```

File: qiskit/qobj/converters/pulse_instruction.py [hunk 1]
```diff
@@ -322,180 +316,327 @@ def convert_bundled_acquires(
...
-            "t0": shift + instruction.start_time,
+            "t0": time_offset + instruction.start_time,
             "ch": instruction.channel.name,
             "frequency": instruction.frequency / 1e9,
         }
         return self._qobj_model(**command_dict)
 
-    @bind_instruction(instructions.SetPhase)
-    def convert_set_phase(self, shift, instruction):
+    def _convert_SetPhase(
```
Comment (mtreinish) [inline qiskit/qobj/converters/pulse_instruction.py:363]: I'm surprised pylint doesn't complain about this being camel case.

**Review by mtreinish — DISMISSED** [review]
Overall this looks good to me, I didn't see anything that really stood out as an issue in reviewing this. However, I'm not the most familiar with this part of the code base. I've not really had any reason to work with pulse qobj converters directly before. I left some review comments inline on the release notes, but they're not really blocking (we can always fix that as part of the release prep PR)

File: qiskit/qobj/converters/pulse_instruction.py (hunk 1 above)
Comment (nkanazawa1989) [reply qiskit/qobj/converters/pulse_instruction.py]: Yes, this was what I wanted to change. Now we can use method dispatch in terra, I updated the implementation of these method in [f1ad415](https://github.com/Qiskit/qiskit-terra/pull/8839/commits/f1ad41575d63c1b99f1f81b5276125fe4d347c6c)

File: qiskit/qobj/converters/pulse_instruction.py [hunk 2]
```diff
@@ -88,96 +92,98 @@ def to_type(cls, name: str) -> library.SymbolicPulse:
...
+        instruction,
+        time_offset: int,
+    ) -> PulseQobjInstruction:
+        raise QiskitError(
+            f"Pulse Qobj doesn't support {instruction.__class__.__name__}. "
+            "This instruction cannot be submitted with Qobj."
+        )
+
+    @_convert_instruction.register(instructions.Acquire)
+    def _(
```
[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit/qobj/converters/pulse_instruction.py]: Fair enough. Done in [392324a](https://github.com/Qiskit/qiskit-terra/pull/8839/commits/392324a2913b11f18db589818916652b51ec4659)

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for the quick update.

**Pull request merged**